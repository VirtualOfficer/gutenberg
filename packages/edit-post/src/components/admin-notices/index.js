/**
 * WordPress dependencies
 */
import { Component } from '@wordpress/element';
import { withDispatch } from '@wordpress/data';

/**
 * Set of class name suffixes to associate as the status of an admin notice.
 *
 * @type {Set}
 */
const NOTICE_STATUSES = new Set( [
	'success',
	'warning',
	'error',
	'info',
] );

/**
 * Pattern matching a class list item matching the pattern of an admin notice
 * status class.
 *
 * @type {RegExp}
 */
const REGEXP_NOTICE_STATUS = /^notice-(\w+)$/;

/**
 * Returns an array of admin notice Elements.
 *
 * @return {Element[]} Admin notice elements.
 */
function getAdminNotices() {
	return [ ...document.querySelectorAll( '.notice' ) ];
}

/**
 * Given an admin notice Element, returns the upgraded content string. Prefers
 * text found within its first child paragraph node, but falls back to the text
 * content of the entire notice element.
 *
 * @param {Element} element Admin notice element.
 *
 * @return {string} Upgraded notice content.
 */
function getNoticeContentFromElement( element ) {
	let node = element;
	if ( node.firstElementChild && node.firstElementChild.nodeName === 'P' ) {
		node = node.firstElementChild;
	}

	return node.textContent.trim();
}

/**
 * Given an admin notice Element, returns the upgraded status type, or
 * undefined if one cannot be determined (i.e. one is not assigned).
 *
 * @param {Element} element Admin notice element.
 *
 * @return {?string} Upgraded status type.
 */
function getNoticeStatusFromClassList( element ) {
	for ( const className of element.classList ) {
		const match = className.match( REGEXP_NOTICE_STATUS );
		if ( match && NOTICE_STATUSES.has( match[ 1 ] ) ) {
			return match[ 1 ];
		}
	}
}

/**
 * Given an admin notice Element, returns a notices module object.
 *
 * @param {Element} element Admin notice element.
 *
 * @return {WPNotice} Notice object.
 */
function getNoticeFromElement( element ) {
	const status = getNoticeStatusFromClassList( element );
	const content = getNoticeContentFromElement( element );
	const isDismissible = element.classList.contains( 'is-dismissible' );

	return { status, content, isDismissible };
}

export class AdminNotices extends Component {
	componentDidMount() {
		this.convertNotices();
	}

	convertNotices() {
		const { createNotice } = this.props;
		getAdminNotices().forEach( ( element ) => {
			// Convert and create.
			const notice = getNoticeFromElement( element );
			createNotice( notice );

			// Remove (now-redundant) admin notice element.
			element.parentNode.removeChild( element );
		} );
	}

	render() {
		return null;
	}
}

export default withDispatch( ( dispatch ) => {
	const { createNotice } = dispatch( 'core/notices' );

	return { createNotice };
} )( AdminNotices );
