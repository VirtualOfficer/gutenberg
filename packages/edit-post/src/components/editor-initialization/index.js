/**
 * WordPress dependencies
 */
import { useEffect, useContext } from '@wordpress/element';
import { RegistryContext } from '@wordpress/data';

/**
 * Internal dependencies
 */
import {
	blockSelectionListener,
	adjustSidebarListener,
	viewPostLinkUpdateListener,
} from './listeners';

/**
 * Data component used for initializing the editor and re-initializes
 * when postId changes or unmount.
 *
 * @param {number} postId
 * @return {null} This is a data component so does not render any ui.
 */
export default function( { postId } ) {
	const registry = useContext( RegistryContext );
	useEffect( () => {
		const blockSelectionUnsubscribe = registry.subscribe(
			blockSelectionListener( registry )
		);
		const adjustSidebarUnsubscribe = registry.subscribe(
			adjustSidebarListener( registry )
		);
		const viewPostLinkUpdateUnsubscribe = registry.subscribe(
			viewPostLinkUpdateListener( registry )
		);
		registry.dispatch( 'core/nux' ).triggerGuide( [
			'core/editor.inserter',
			'core/editor.settings',
			'core/editor.preview',
			'core/editor.publish',
		] );
		return () => {
			blockSelectionUnsubscribe();
			adjustSidebarUnsubscribe();
			viewPostLinkUpdateUnsubscribe();
		};
	}, [ registry, postId ] );
	return null;
}
