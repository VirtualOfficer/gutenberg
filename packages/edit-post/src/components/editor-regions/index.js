/**
 * External dependencies
 */
import classnames from 'classnames';

function EditorRegions( { footer, header, sidebar, content, className } ) {
	return (
		<div className={ classnames( className, 'edit-post-editor-regions' ) }>
			{ !! header && (
				<div className="edit-post-editor-regions__header">
					{ header }
				</div>
			) }
			<div className="edit-post-editor-regions__body">
				<div className="edit-post-editor-regions__content">{ content }</div>
				{ !! sidebar && (
					<div className="edit-post-editor-regions__sidebar">
						{ sidebar }
					</div>
				) }
			</div>
			{ !! footer && (
				<div className="edit-post-editor-regions__footer">
					{ footer }
				</div>
			) }
		</div>
	);
}

export default EditorRegions;
