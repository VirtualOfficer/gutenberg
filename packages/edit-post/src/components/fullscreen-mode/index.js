/**
 * WordPress dependencies
 */
import { Component } from '@wordpress/element';
import { withSelect } from '@wordpress/data';
import { doAction } from '@wordpress/hooks';

class FullscreenMode extends Component {
	componentDidMount() {
		this.sync();
	}

	componentDidUpdate( prevProps ) {
		if ( this.props.isActive !== prevProps.isActive ) {
			this.sync();
		}
	}

	sync() {
		const { isActive } = this.props;
		let delay = 0;
		if ( isActive ) {
			document.body.classList.add( 'is-fullscreen-mode' );
			delay = 300; // .is-fullscreen-mode has a fade in animation of 0.3s
		} else {
			document.body.classList.remove( 'is-fullscreen-mode' );
		}

		setTimeout( () => doAction( 'fullscreenModeToggled' ), delay );
	}

	render() {
		return null;
	}
}

export default withSelect( ( select ) => ( {
	isActive: select( 'core/edit-post' ).isFeatureActive( 'fullscreenMode' ),
} ) )( FullscreenMode );
