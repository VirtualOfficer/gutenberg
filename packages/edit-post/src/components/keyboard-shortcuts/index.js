/**
 * WordPress dependencies
 */
import { useEffect } from '@wordpress/element';
import { useDispatch } from '@wordpress/data';
import {
	useShortcut,
	store as keyboardShortcutsStore,
} from '@wordpress/keyboard-shortcuts';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { store as editPostStore } from '../../store';

function KeyboardShortcuts() {
	const { toggleFeature } = useDispatch( editPostStore );
	const { registerShortcut } = useDispatch( keyboardShortcutsStore );

	useEffect( () => {
		registerShortcut( {
			name: 'core/edit-post/toggle-fullscreen',
			category: 'global',
			description: __( 'Toggle fullscreen mode.' ),
			keyCombination: {
				modifier: 'secondary',
				character: 'f',
			},
		} );

		registerShortcut( {
			name: 'core/edit-post/next-region',
			category: 'global',
			description: __( 'Navigate to the next part of the editor.' ),
			keyCombination: {
				modifier: 'ctrl',
				character: '`',
			},
			aliases: [
				{
					modifier: 'access',
					character: 'n',
				},
			],
		} );

		registerShortcut( {
			name: 'core/edit-post/previous-region',
			category: 'global',
			description: __( 'Navigate to the previous part of the editor.' ),
			keyCombination: {
				modifier: 'ctrlShift',
				character: '`',
			},
			aliases: [
				{
					modifier: 'access',
					character: 'p',
				},
				{
					modifier: 'ctrlShift',
					character: '~',
				},
			],
		} );
	}, [] );

	useShortcut( 'core/edit-post/toggle-fullscreen', () => {
		toggleFeature( 'fullscreenMode' );
	} );

	return null;
}

export default KeyboardShortcuts;
