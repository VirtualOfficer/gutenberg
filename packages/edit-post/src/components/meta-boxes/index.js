/**
 * External dependencies
 */
import { map } from 'lodash';

/**
 * WordPress dependencies
 */
import { useSelect, useRegistry } from '@wordpress/data';
import { useEffect } from '@wordpress/element';
import { store as editorStore } from '@wordpress/editor';

/**
 * Internal dependencies
 */
import MetaBoxesArea from './meta-boxes-area';
import MetaBoxVisibility from './meta-box-visibility';
import { store as editPostStore } from '../../store';

export default function MetaBoxes( { location } ) {
	const registry = useRegistry();
	const {
		metaBoxes,
		isVisible,
		metaBoxesInitialized,
		isEditorReady,
	} = useSelect(
		( select ) => {
			const { __unstableIsEditorReady } = select( editorStore );
			const {
				isMetaBoxLocationVisible,
				getMetaBoxesPerLocation,
				metaBoxesInitialized: _metaBoxesInitialized,
			} = select( editPostStore );
			return {
				metaBoxes: getMetaBoxesPerLocation( location ),
				isVisible: isMetaBoxLocationVisible( location ),
				metaBoxesInitialized: _metaBoxesInitialized(),
				isEditorReady: __unstableIsEditorReady(),
			};
		},
		[ location ]
	);

	// When editor is ready, initialize postboxes (wp core script) and metabox
	// saving. This initializes all meta box locations, not just this specific
	// one.
	useEffect( () => {
		if ( isEditorReady ) {
			registry.dispatch( editPostStore ).initializeMetaBoxes();
		}
	}, [ isEditorReady ] );

	if ( ! metaBoxesInitialized ) {
		return null;
	}

	return (
		<>
			{ map( metaBoxes, ( { id } ) => (
				<MetaBoxVisibility key={ id } id={ id } />
			) ) }
			{ isVisible && <MetaBoxesArea location={ location } /> }
		</>
	);
}
