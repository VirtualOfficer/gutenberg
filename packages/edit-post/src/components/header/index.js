/**
 * External Dependencies
 */
import { get } from 'lodash';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { IconButton } from '@wordpress/components';
import {
	PostPreviewButton,
	PostSavedState,
	PostPublishPanelToggle,
	PostPublishButton,
} from '@wordpress/editor';
import { withDispatch, withSelect } from '@wordpress/data';
import { compose } from '@wordpress/compose';
import { DotTip } from '@wordpress/nux';

/**
 * Internal dependencies
 */
import MoreMenu from './more-menu';
import HeaderToolbar from './header-toolbar';
import PinnedPlugins from './pinned-plugins';
import shortcuts from '../../keyboard-shortcuts';

function Header( {
	isEditorSidebarOpened,
	openGeneralSidebar,
	closeGeneralSidebar,
	isPublishSidebarOpened,
	isPublishSidebarEnabled,
	togglePublishSidebar,
	hasActiveMetaboxes,
	hasPublishAction,
	isSaving,
	isPublished,
	isScheduled,
	isBeingScheduled,
	isPending,
} ) {
	const toggleGeneralSidebar = isEditorSidebarOpened ? closeGeneralSidebar : openGeneralSidebar;

	const shouldShowButton = (
		! isPublishSidebarEnabled ||
		isPublished ||
		( isScheduled && isBeingScheduled ) ||
		( isPending && ! hasPublishAction )
	);
	return (
		<div
			role="region"
			/* translators: accessibility text for the top bar landmark region. */
			aria-label={ __( 'Editor top bar' ) }
			className="edit-post-header"
			tabIndex="-1"
		>
			<HeaderToolbar />
			{ ! isPublishSidebarOpened && (
				<div className="edit-post-header__settings">
					<PostSavedState
						forceIsDirty={ hasActiveMetaboxes }
						forceIsSaving={ isSaving }
					/>
					<PostPreviewButton />
					{ shouldShowButton ? (
						<PostPublishButton
							forceIsDirty={ hasActiveMetaboxes }
							forceIsSaving={ isSaving }
						/>
					) : (
						<PostPublishPanelToggle
							isOpen={ isPublishSidebarOpened }
							onToggle={ togglePublishSidebar }
							forceIsDirty={ hasActiveMetaboxes }
							forceIsSaving={ isSaving }
						/>
					) }
					<div>
						<IconButton
							icon="admin-generic"
							label={ __( 'Settings' ) }
							onClick={ toggleGeneralSidebar }
							isToggled={ isEditorSidebarOpened }
							aria-expanded={ isEditorSidebarOpened }
							shortcut={ shortcuts.toggleSidebar }
						/>
						<DotTip id="core/editor.settings">
							{ __( 'You’ll find more settings for your page and blocks in the sidebar. Click “Settings” to open it.' ) }
						</DotTip>
					</div>
					<PinnedPlugins.Slot />
					<MoreMenu />
				</div>
			) }
		</div>
	);
}

export default compose(
	withSelect( ( select ) => ( {
		isEditorSidebarOpened: select( 'core/edit-post' ).isEditorSidebarOpened(),
		isPublishSidebarOpened: select( 'core/edit-post' ).isPublishSidebarOpened(),
		isPublishSidebarEnabled: select( 'core/editor' ).isPublishSidebarEnabled(),
		hasActiveMetaboxes: select( 'core/edit-post' ).hasMetaBoxes(),
		isSaving: select( 'core/edit-post' ).isSavingMetaBoxes(),
		hasBlockSelection: !! select( 'core/editor' ).getBlockSelectionStart(),
		hasPublishAction: get( select( 'core/editor' ).getCurrentPost(), [ '_links', 'wp:action-publish' ], false ),
		isPublished: select( 'core/editor' ).isCurrentPostPublished(),
		isScheduled: select( 'core/editor' ).isCurrentPostScheduled(),
		isBeingScheduled: select( 'core/editor' ).isEditedPostBeingScheduled(),
		isPending: select( 'core/editor' ).isCurrentPostPending(),
	} ) ),
	withDispatch( ( dispatch, { hasBlockSelection } ) => {
		const { openGeneralSidebar, closeGeneralSidebar, togglePublishSidebar } = dispatch( 'core/edit-post' );
		const sidebarToOpen = hasBlockSelection ? 'edit-post/block' : 'edit-post/document';
		return {
			openGeneralSidebar: () => openGeneralSidebar( sidebarToOpen ),
			closeGeneralSidebar: closeGeneralSidebar,
			togglePublishSidebar: togglePublishSidebar,
			hasBlockSelection: undefined,
		};
	} ),
)( Header );
