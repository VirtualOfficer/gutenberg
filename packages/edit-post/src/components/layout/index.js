/**
 * External dependencies
 */
import clsx from 'clsx';

/**
 * WordPress dependencies
 */
import {
	AutosaveMonitor,
	LocalAutosaveMonitor,
	UnsavedChangesWarning,
	EditorNotices,
	EditorKeyboardShortcutsRegister,
	EditorKeyboardShortcuts,
	EditorSnackbars,
	store as editorStore,
	privateApis as editorPrivateApis,
} from '@wordpress/editor';
import { useSelect, useDispatch } from '@wordpress/data';
import {
	BlockBreadcrumb,
	BlockToolbar,
	privateApis as blockEditorPrivateApis,
	store as blockEditorStore,
} from '@wordpress/block-editor';
import { ScrollLock } from '@wordpress/components';
import { useViewportMatch } from '@wordpress/compose';
import { PluginArea } from '@wordpress/plugins';
import { __, _x, sprintf } from '@wordpress/i18n';
import { useState, useCallback, useMemo } from '@wordpress/element';
import { store as keyboardShortcutsStore } from '@wordpress/keyboard-shortcuts';
import { store as noticesStore } from '@wordpress/notices';
import { store as preferencesStore } from '@wordpress/preferences';
import { privateApis as commandsPrivateApis } from '@wordpress/commands';
import { privateApis as coreCommandsPrivateApis } from '@wordpress/core-commands';
import { privateApis as blockLibraryPrivateApis } from '@wordpress/block-library';
import { addQueryArgs } from '@wordpress/url';

/**
 * Internal dependencies
 */
import TextEditor from '../text-editor';
import VisualEditor from '../visual-editor';
import EditPostKeyboardShortcuts from '../keyboard-shortcuts';
import InitPatternModal from '../init-pattern-modal';
import BrowserURL from '../browser-url';
import Header from '../header';
import MetaBoxes from '../meta-boxes';
import WelcomeGuide from '../welcome-guide';
import { store as editPostStore } from '../../store';
import { unlock } from '../../lock-unlock';
import useCommonCommands from '../../hooks/commands/use-common-commands';

const { getLayoutStyles } = unlock( blockEditorPrivateApis );
const { useCommands } = unlock( coreCommandsPrivateApis );
const { useCommandContext } = unlock( commandsPrivateApis );
const {
	InserterSidebar,
	ListViewSidebar,
	ComplementaryArea,
	FullscreenMode,
	SavePublishPanels,
	InterfaceSkeleton,
	interfaceStore,
	Sidebar,
} = unlock( editorPrivateApis );
const { BlockKeyboardShortcuts } = unlock( blockLibraryPrivateApis );

const interfaceLabels = {
	/* translators: accessibility text for the editor top bar landmark region. */
	header: __( 'Editor top bar' ),
	/* translators: accessibility text for the editor content landmark region. */
	body: __( 'Editor content' ),
	/* translators: accessibility text for the editor settings landmark region. */
	sidebar: __( 'Editor settings' ),
	/* translators: accessibility text for the editor publish landmark region. */
	actions: __( 'Editor publish' ),
	/* translators: accessibility text for the editor footer landmark region. */
	footer: __( 'Editor footer' ),
};

function useEditorStyles() {
	const { hasThemeStyleSupport, editorSettings } = useSelect(
		( select ) => ( {
			hasThemeStyleSupport:
				select( editPostStore ).isFeatureActive( 'themeStyles' ),
			editorSettings: select( editorStore ).getEditorSettings(),
		} ),
		[]
	);

	// Compute the default styles.
	return useMemo( () => {
		const presetStyles =
			editorSettings.styles?.filter(
				( style ) =>
					style.__unstableType && style.__unstableType !== 'theme'
			) ?? [];

		const defaultEditorStyles = [
			...editorSettings.defaultEditorStyles,
			...presetStyles,
		];

		// Has theme styles if the theme supports them and if some styles were not preset styles (in which case they're theme styles).
		const hasThemeStyles =
			hasThemeStyleSupport &&
			presetStyles.length !== ( editorSettings.styles?.length ?? 0 );

		// If theme styles are not present or displayed, ensure that
		// base layout styles are still present in the editor.
		if ( ! editorSettings.disableLayoutStyles && ! hasThemeStyles ) {
			defaultEditorStyles.push( {
				css: getLayoutStyles( {
					style: {},
					selector: 'body',
					hasBlockGapSupport: false,
					hasFallbackGapSupport: true,
					fallbackGapValue: '0.5em',
				} ),
			} );
		}

		return hasThemeStyles ? editorSettings.styles : defaultEditorStyles;
	}, [
		editorSettings.defaultEditorStyles,
		editorSettings.disableLayoutStyles,
		editorSettings.styles,
		hasThemeStyleSupport,
	] );
}

function Layout( { initialPost } ) {
	useCommands();
	useCommonCommands();

	const isMobileViewport = useViewportMatch( 'medium', '<' );
	const isWideViewport = useViewportMatch( 'large' );
	const isLargeViewport = useViewportMatch( 'medium' );

	const { closeGeneralSidebar } = useDispatch( editPostStore );
	const { createErrorNotice } = useDispatch( noticesStore );
	const {
		mode,
		isFullscreenActive,
		isRichEditingEnabled,
		sidebarIsOpened,
		hasActiveMetaboxes,
		previousShortcut,
		nextShortcut,
		hasBlockSelected,
		isInserterOpened,
		isListViewOpened,
		showIconLabels,
		isDistractionFree,
		showBlockBreadcrumbs,
		showMetaBoxes,
		documentLabel,
		hasHistory,
		hasBlockBreadcrumbs,
		blockEditorMode,
		isEditingTemplate,
	} = useSelect( ( select ) => {
		const { get } = select( preferencesStore );
		const { getEditorSettings, getPostTypeLabel } = select( editorStore );
		const editorSettings = getEditorSettings();
		const postTypeLabel = getPostTypeLabel();

		return {
			showMetaBoxes:
				select( editorStore ).getRenderingMode() === 'post-only',
			sidebarIsOpened: !! (
				select( interfaceStore ).getActiveComplementaryArea( 'core' ) ||
				select( editorStore ).isPublishSidebarOpened()
			),
			isFullscreenActive:
				select( editPostStore ).isFeatureActive( 'fullscreenMode' ),
			isInserterOpened: select( editorStore ).isInserterOpened(),
			isListViewOpened: select( editorStore ).isListViewOpened(),
			mode: select( editorStore ).getEditorMode(),
			isRichEditingEnabled: editorSettings.richEditingEnabled,
			hasActiveMetaboxes: select( editPostStore ).hasMetaBoxes(),
			previousShortcut: select(
				keyboardShortcutsStore
			).getAllShortcutKeyCombinations( 'core/editor/previous-region' ),
			nextShortcut: select(
				keyboardShortcutsStore
			).getAllShortcutKeyCombinations( 'core/editor/next-region' ),
			showIconLabels: get( 'core', 'showIconLabels' ),
			isDistractionFree: get( 'core', 'distractionFree' ),
			showBlockBreadcrumbs: get( 'core', 'showBlockBreadcrumbs' ),
			// translators: Default label for the Document in the Block Breadcrumb.
			documentLabel: postTypeLabel || _x( 'Document', 'noun' ),
			hasBlockSelected:
				!! select( blockEditorStore ).getBlockSelectionStart(),
			hasHistory: !! getEditorSettings().onNavigateToPreviousEntityRecord,
			hasBlockBreadcrumbs: get( 'core', 'showBlockBreadcrumbs' ),
			blockEditorMode:
				select( blockEditorStore ).__unstableGetEditorMode(),
			isEditingTemplate:
				select( editorStore ).getCurrentPostType() === 'wp_template',
		};
	}, [] );

	// Set the right context for the command palette
	const commandContext = hasBlockSelected
		? 'block-selection-edit'
		: 'post-editor-edit';
	useCommandContext( commandContext );

	const styles = useEditorStyles();

	// Local state for save panel.
	// Note 'truthy' callback implies an open panel.
	const [ entitiesSavedStatesCallback, setEntitiesSavedStatesCallback ] =
		useState( false );

	const closeEntitiesSavedStates = useCallback(
		( arg ) => {
			if ( typeof entitiesSavedStatesCallback === 'function' ) {
				entitiesSavedStatesCallback( arg );
			}
			setEntitiesSavedStatesCallback( false );
		},
		[ entitiesSavedStatesCallback ]
	);

	// We need to add the show-icon-labels class to the body element so it is applied to modals.
	if ( showIconLabels ) {
		document.body.classList.add( 'show-icon-labels' );
	} else {
		document.body.classList.remove( 'show-icon-labels' );
	}

	const className = clsx( 'edit-post-layout', 'is-mode-' + mode, {
		'is-sidebar-opened': sidebarIsOpened,
		'has-metaboxes': hasActiveMetaboxes,
		'is-distraction-free': isDistractionFree && isWideViewport,
		'is-entity-save-view-open': !! entitiesSavedStatesCallback,
		'has-block-breadcrumbs':
			hasBlockBreadcrumbs && ! isDistractionFree && isWideViewport,
	} );

	const secondarySidebarLabel = isListViewOpened
		? __( 'Document Overview' )
		: __( 'Block Library' );

	const secondarySidebar = () => {
		if ( mode === 'visual' && isInserterOpened ) {
			return (
				<InserterSidebar
					closeGeneralSidebar={ closeGeneralSidebar }
					isRightSidebarOpen={ sidebarIsOpened }
				/>
			);
		}
		if ( mode === 'visual' && isListViewOpened ) {
			return <ListViewSidebar />;
		}

		return null;
	};

	function onPluginAreaError( name ) {
		createErrorNotice(
			sprintf(
				/* translators: %s: plugin name */
				__(
					'The "%s" plugin has encountered an error and cannot be rendered.'
				),
				name
			)
		);
	}

	const { createSuccessNotice } = useDispatch( noticesStore );

	const onActionPerformed = useCallback(
		( actionId, items ) => {
			switch ( actionId ) {
				case 'move-to-trash':
					{
						document.location.href = addQueryArgs( 'edit.php', {
							trashed: 1,
							post_type: items[ 0 ].type,
							ids: items[ 0 ].id,
						} );
					}
					break;
				case 'duplicate-post':
					{
						const newItem = items[ 0 ];
						const title =
							typeof newItem.title === 'string'
								? newItem.title
								: newItem.title?.rendered;
						createSuccessNotice(
							sprintf(
								// translators: %s: Title of the created post e.g: "Post 1".
								__( '"%s" successfully created.' ),
								title
							),
							{
								type: 'snackbar',
								id: 'duplicate-post-action',
								actions: [
									{
										label: __( 'Edit' ),
										onClick: () => {
											const postId = newItem.id;
											document.location.href =
												addQueryArgs( 'post.php', {
													post: postId,
													action: 'edit',
												} );
										},
									},
								],
							}
						);
					}
					break;
			}
		},
		[ createSuccessNotice ]
	);

	return (
		<>
			<FullscreenMode isActive={ isFullscreenActive } />
			<BrowserURL hasHistory={ hasHistory } />
			<UnsavedChangesWarning />
			<AutosaveMonitor />
			<LocalAutosaveMonitor />
			<EditPostKeyboardShortcuts />
			<EditorKeyboardShortcutsRegister />
			<EditorKeyboardShortcuts />
			<BlockKeyboardShortcuts />

			<InterfaceSkeleton
				isDistractionFree={ isDistractionFree && isWideViewport }
				className={ className }
				labels={ {
					...interfaceLabels,
					secondarySidebar: secondarySidebarLabel,
				} }
				header={
					<Header
						setEntitiesSavedStatesCallback={
							setEntitiesSavedStatesCallback
						}
						initialPost={ initialPost }
					/>
				}
				editorNotices={ <EditorNotices /> }
				secondarySidebar={ secondarySidebar() }
				sidebar={
					! isDistractionFree && (
						<ComplementaryArea.Slot scope="core" />
					)
				}
				notices={ <EditorSnackbars /> }
				content={
					<>
						{ ! isDistractionFree && <EditorNotices /> }
						{ ( mode === 'text' || ! isRichEditingEnabled ) && (
							<TextEditor />
						) }
						{ ! isLargeViewport && <BlockToolbar hideDragHandle /> }
						{ isRichEditingEnabled && mode === 'visual' && (
							<VisualEditor styles={ styles } />
						) }
						{ ! isDistractionFree && showMetaBoxes && (
							<div className="edit-post-layout__metaboxes">
								<MetaBoxes location="normal" />
								<MetaBoxes location="advanced" />
							</div>
						) }
						{ isMobileViewport && sidebarIsOpened && (
							<ScrollLock />
						) }
					</>
				}
				footer={
					! isDistractionFree &&
					! isMobileViewport &&
					showBlockBreadcrumbs &&
					isRichEditingEnabled &&
					blockEditorMode !== 'zoom-out' &&
					mode === 'visual' && (
						<div className="edit-post-layout__footer">
							<BlockBreadcrumb rootLabelText={ documentLabel } />
						</div>
					)
				}
				actions={
					<SavePublishPanels
						closeEntitiesSavedStates={ closeEntitiesSavedStates }
						isEntitiesSavedStatesOpen={
							entitiesSavedStatesCallback
						}
						setEntitiesSavedStatesCallback={
							setEntitiesSavedStatesCallback
						}
						forceIsDirtyPublishPanel={ hasActiveMetaboxes }
					/>
				}
				shortcuts={ {
					previous: previousShortcut,
					next: nextShortcut,
				} }
			/>
			<WelcomeGuide />
			<InitPatternModal />
			<PluginArea onError={ onPluginAreaError } />
			{ ! isDistractionFree && (
				<Sidebar
					onActionPerformed={ onActionPerformed }
					extraPanels={
						! isEditingTemplate && <MetaBoxes location="side" />
					}
				/>
			) }
		</>
	);
}

export default Layout;
