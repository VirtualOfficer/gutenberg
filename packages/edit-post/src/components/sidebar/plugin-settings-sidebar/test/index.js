/**
 * WordPress dependencies
 */
import { SlotFillProvider } from '@wordpress/components';
import { render } from '@wordpress/element';

/**
 * Internal dependencies
 */
import PluginSettingsSidebar from '../';

jest.mock( '../../../../../../components/src/button' );

describe( 'PluginSettingsSidebar', () => {
	test( 'renders fill properly', () => {
		const div = document.createElement( 'div' );
		render(
			<SlotFillProvider>
				<PluginSettingsSidebar>
					My panel content
				</PluginSettingsSidebar>
				<PluginSettingsSidebar.Slot />
			</SlotFillProvider>,
			div
		);

		expect( div.innerHTML ).toMatchSnapshot();
	} );
} );
