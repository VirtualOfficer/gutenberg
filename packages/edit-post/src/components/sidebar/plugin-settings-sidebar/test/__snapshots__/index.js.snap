// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginSettingsSidebar renders fill properly 1`] = `"My panel content"`;
